Shared/configurations/ConfigurationController.cpp

```diff
@@ -64,26 +79,183 @@ void ConfigurationController::select(int index)
+  if (rootDataDirectoryVar.isUndefined())
+    return false;
+
+  // strip off any trailing slashes if they exist
+  auto rootDataDirectoryCleaned = rootDataDirectoryVar.toString();
+  while (rootDataDirectoryCleaned.back() == '\\' || rootDataDirectoryCleaned.back() == '/')
+    rootDataDirectoryCleaned.chop(1);
+
+  // overwrite the app config file with the new json doc
+  if (!dsaAppConfigFile.open(QIODevice::WriteOnly | QIODevice::Truncate))
+    return false;
+
+  // re-write the entire document with the old root directory replaced by the extracted location
+  QString configString{configBytes};
+  const QString rootDataDirectoryNew{configurationDirectory.absolutePath()};
+  const QString configStringUpdated{configString.replace(rootDataDirectoryCleaned, rootDataDirectoryNew)};
+  dsaAppConfigFile.write(configStringUpdated.toUtf8());
+  return true;
+}
+
+void ConfigurationController::resetConfigurationDeviceStatus(const QString& configurationName)
+{
+  if (!m_configurationListModel->setDataByName(configurationName, 0, ConfigurationListModel::PercentDownloaded) ||
+      !m_configurationListModel->setDataByName(configurationName, 0, ConfigurationListModel::PercentExtracted))
+    emit toolErrorOccurred(QStringLiteral("Error resetting the configurations status on the device"), QStringLiteral("Configuration Error"));
+}
+
 void ConfigurationController::download(int index)
 {
-  // prevent calling the download function if something is already in progress
-  if (m_downloadInProgress)
+  // fetch a copy of the selected model item from the listview
+  const auto configuration = m_configurationListModel->at(index);
+  const auto configurationUrl = configuration.url();
+  const auto configurationName = configuration.name();
+  if (configurationUrl.isLocalFile())
+  {
+    m_configurationListModel->setDataByName(configurationName, 100, ConfigurationListModel::PercentDownloaded);
+    extractConfigurationDownload(configurationUrl.toLocalFile(), configurationName);
     return;
+  }
 
-  // get the url from the selected item
-  auto i = m_configurationListModel->index(index, 0);
-  auto url = m_configurationListModel->data(i, ConfigurationListModel::Roles::Url);
-  m_activeDownloadIndex = index;
+  // attempt to download the data as a portal item if it matches the item id url pattern
+  const auto configurationUrlStr = configuration.urlStr();
+  static QRegularExpression regexPortalItem{REGEX_PORTAL_ITEM_URL};
+  if (regexPortalItem.match(configurationUrlStr).hasMatch())
+  {
+    PortalItem checkPortalItem{configurationUrl}; // using a local portal item here to parse the portal and item id out of the url
+    auto* portal = new Portal(checkPortalItem.portal()->url(), this);
+    auto* portalItem = new PortalItem(portal, checkPortalItem.itemId(), portal);
+    connect(portalItem, &PortalItem::doneLoading, this, [this, portal, portalItem, configurationName](const Error& error)
```

Comment:
This is optional, but if there are large lambdas, consider splitting them out into their own methods. For example `onPortalItemLoaded_` and then this connection can just call that directly.

The `download` method looks a bit large with several connection inside of it.